File: Deblurring-Matlab/P4IP.m
function [x] = P4IP(y,beta,lambda,lambda_step,shape,H,max_iter,epsilon,verbose) 
%P4IP Summary of this function goes here
%   Detailed explanation goes here
u = zeros(shape);
nnn = 0.002*randn(size(y));
x = reshape(H(reshape(H(y),shape) + nnn.*(nnn>=0)),shape);
x = x;

v = x;

figure("Name","Initial guess");
imshow(x/max(max(x)));

for k= 1:max_iter
    x_prev = x;
    u_prev = u;
    v_prev = v;
    temp_x = ADMM(y(:),lambda,v_prev(:),u_prev(:),x_prev(:),H,verbose);
    x = reshape(temp_x,shape);
    v = BM3D(x + u_prev,sqrt(beta/lambda),'np',BM3DProfile.ALL_STAGES);
    u = u_prev + x - v;
    
    lambda = lambda*lambda_step;
    
    if verbose == 1 || rem(k,10) == 0
        fprintf("Iteration %d : RRMSE %f\n",k,rrmse(x_prev,x,1));
    end

    if rrmse(x_prev,x,1) < epsilon
        fprintf("Breaking due to epsilon\n");
        break
    end
end
 
end

File: Python/P4IP.py
from types import LambdaType
import bm3d
import numpy as np
import cv2
from ADMM import ADMM

class P4IP:

    def __init__(self,y, beta, Lambda, lambdaStep, shape, H = None,verbose=0):
        self.H = H
        self.y = y.ravel()
        self.shape = shape
        self.beta = beta
        self.Lambda = Lambda
        self.lambdaStep = lambdaStep
        self.verbose = verbose

    def zero_init(self, iter = 100):
        self.u_0 = np.zeros(self.shape)
        self.v_0 = self.y.reshape(self.shape)
        self.iter = iter
    
    def reconstruct(self):
        u = self.u_0
        v = self.v_0
        x = self.y.reshape(self.shape)
        Lambda = self.Lambda
        
        if self.verbose:
            print(v.shape,u.shape,self.y.shape)
        for k in range(self.iter):
            x = ADMM(self.y.copy(),Lambda,v.copy(),u.copy(),x.copy()+0.1*np.random.randn(*x.shape),self.H) 
            v = bm3d.bm3d(x + u, stage_arg=bm3d.BM3DStages.ALL_STAGES, sigma_psd = np.sqrt(self.beta / self.Lambda))
            u = u + x - v
            Lambda = Lambda*self.lambdaStep
            print(k)
        
        return x

            


